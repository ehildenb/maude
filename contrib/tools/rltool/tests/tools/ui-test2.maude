set include BOOL off .

fmod CHOICE-DATA is
  sorts Nat MSet State Pred .
  subsorts Nat < MSet .
  op zero : -> Nat [ctor] .
  op s    : Nat -> Nat [ctor] .
  op __   : MSet MSet -> MSet [ctor assoc comm] .
  op {_}  : MSet -> State [ctor] .
  op tt   : -> Pred [ctor] .
  op _=C_ : MSet MSet -> Pred [ctor] .  *** MSet containment
  vars U V : MSet . var N : Nat .
  eq U =C U = tt .
  eq U =C U V = tt .
endfm

mod CHOICE is pr CHOICE-DATA .
  vars U V : MSet .
  --- add extra bracket to test invariant creation
  op [_] : MSet -> State .
  rl [choice] : {U V} => {U} .
endm

load ../../rltool.maude

set show stats off .
set show timing off .
set show loop stats off .
set show loop timing off .
set print conceal on .
print conceal mod_is_sorts_._____endm fmod_is_sorts_.____endfm .

(select CHOICE .)
(use tool varsat for validity on CHOICE-DATA .)
(declare-vars (M:MSet) U (N:Nat) .)
(def-term-set ([N]) | true .)
(inv test-invariant-1 to '`[_`] on ({M}) | true .)
(inv test-invariant-2 to '`[_`] on ({N}) | true .)
(start-proof .)
--- these splits will succeed, since they do not change
--- the sharing of any variable
---set break on .
break select split-action .
(split 1 by (M&1:MSet) = (N) .)
(replace 4 by (M&1:MSet) =/= (N':Nat) .)
(split 5 by (N':Nat) = (N':Nat) and (M&1:MSet) = (N':Nat) .)
--- these splits will fail due to to oversharing
(split 2 by (M:MSet) = (N&2:Nat) .)
(replace 3 by (M&2:MSet) =/= (N':Nat) .)
(split 6 by (N':Nat) = (N':Nat) and (M&2:MSet) = (N':Nat) .)
--- TOOD: add some splits that fail due to undersharing
(step .)

--- test command parser against other use tool variants
(use tool varsat for validty on CHOICE-DATA with CHOICE-DATA .)
(use tool varsat for validty on CHOICE-DATA with CHOICE-DATA CHOICE-DATA .)

quit
