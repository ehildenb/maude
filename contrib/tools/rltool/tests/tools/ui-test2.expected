
	    Maude Reachability Logic Prover v0.2.9 -- Sep. 21st 2018

Set module to CHOICE
Loaded function varsat for validity
Declared variable(s):
       { M:MSet, N:Nat }
Added terminating state:
       [N:Nat] |  true
Added goal(s):
       [test-invariant-1 : {M&1:MSet} |  true  => [M&2:MSet] |  true]
Added goal(s):
       [test-invariant-2 : {N&1:Nat} |  true  => [N&2:Nat] |  true]
Started proof:
       [1     | {M&1:MSet} |  true  => [M&2:MSet] |  true]
       [2     | {N&1:Nat} |  true  => [N&2:Nat] |  true]
Split rule generated:
       [3     | {M&1:MSet} | (M&1:MSet = N:Nat /\ true) => [M&2:MSet] |  true]
       [4     | {M&1:MSet} | (M&1:MSet =/= N:Nat /\ true) => [M&2:MSet] | 
    true]
Split rule generated:
       [5     | {M&1:MSet} |  M&1:MSet =/= N':Nat  => [M&2:MSet] |  true]
Split rule generated:
       [6     | {M&1:MSet} |  M&1:MSet = N':Nat  => [M&2:MSet] |  true]
       [7     | {M&1:MSet} |  N':Nat = N':Nat  => [M&2:MSet] |  true]
Warning: split rule formulas have unsafe variable sharing
Notice: proof branch failed: [2     | {N&1:Nat} |  true  => [N&2:Nat] |  true]
Split rule generated:
        
Warning: split rule formulas have unsafe variable sharing
Notice: proof branch failed: [3     | {M&1:MSet} | (M&1:MSet = N:Nat /\ true)
    => [M&2:MSet] |  true]
Split rule generated:
        
Warning: split rule formulas have unsafe variable sharing
Notice: proof branch failed: [6     | {M&1:MSet} |  M&1:MSet = N':Nat  => [
    M&2:MSet] |  true]
Split rule generated:
        
Step Results:
       [15    | {&10:MSet} |  N'&11:Nat = N'&11:Nat  => [M&12:MSet] |  true]
Loaded function varsat for validty
Loaded function varsat for validty
Bye.
